---
title: 'scRNA QC'
author: "FHIL\nDerrikGratz"
date: '`r Sys.Date()`'
output: html_document
---

```{r, include=FALSE}
knitr::opts_chunk$set(echo = FALSE,
                      fig.width = 8, fig.height = 6,
                      cache=FALSE, cache.lazy = FALSE) 

library(here)       ## Easier specification of file locations
source(here('figure_scripts/utils.R'))
set.seed(33)
```

# Load data

```{r}
metadata_3p <- read.csv(here('config/3p/metadata.csv')) |>
  mutate(Kit = factor(Kit, levels = kit_order_3p))
sat_curves <- read.csv(here('data/3p/saturation_curves/manually_downsampled.csv')) |>
  select(-Kit) |>
  mutate(across(all_of(c('median_genes', 'median_umi', 'nreads')), as.numeric))
```

# Prepare plotdata

```{r}
zeroes <- data.frame(Sample=metadata_3p$Sample, nreads=0, median_genes=0, median_umi=0)
sat_curves <- rbind(sat_curves, zeroes)
sat_curves <- merge(sat_curves, metadata_3p, by='Sample', all=TRUE)
```

# Modeling

## Fit model

```{r}
model_function <- function(nreads, a, c) {
  # michaelis menten
  (a * nreads) / (c + nreads)
}
models_g <- lapply(unique(metadata_3p$Kit), function(kit) {
  model_data <- sat_curves |>
    filter(Kit == kit) |>
    select(nreads, median_genes) |>
    na.omit()
  fit <- nls(
    median_genes ~ model_function(nreads, a, c),
    data = model_data,
    start = list(a = 3000, c = 10000) # expected end: 3000 genes, 10000 reads
  )
  predictions <- predict(fit, model_data$nreads)
  error <- model_data$median_genes - predictions
  rmse <- sqrt(mean(error^2))
  r2 <- 1 - (sum(error) / sum((model_data$median_genes - mean(model_data$median_genes))^2))
  return(list(model=fit, rmse=rmse, r2=r2))
})
names(models_g) <- unique(metadata_3p$Kit)

models_u <- lapply(unique(metadata_3p$Kit), function(kit) {
  model_data <- sat_curves |>
    filter(Kit == kit) |>
    select(nreads, median_umi) |>
    na.omit()
  fit <- nls(
    median_umi ~ model_function(nreads, a, c),
    data = model_data,
    start = list(a = 5000, c = 10000) # expected end: 5000 umi, 10000 reads
  )
  predictions <- predict(fit, model_data$nreads)
  error <- model_data$median_umi - predictions
  rmse <- sqrt(mean(error^2))
  r2 <- 1 - (sum(error) / sum((model_data$median_umi - mean(model_data$median_umi))^2))
  return(list(model=fit, rmse=rmse, r2=r2))
})
names(models_u) <- unique(metadata_3p$Kit)
```

## Save models

```{r}
saveRDS(list(models_g = models_g, models_u = models_u), here('rds/3p/sat_models.rds'))
```

```{r}
models <- readRDS(here('rds/3p/sat_models.rds'))
models_u <- models$models_u
models_g <- models$models_g
rm(models)
```


## predict data

```{r}
readrange <- seq(0, 25000, 1000)
predicted_g <- lapply(models_g, function(model) {
  predict(model$model, list(nreads=readrange))
}) |> as.data.table() |>
  mutate(nreads=readrange) |>
  melt(id.vars = 'nreads', variable.name = 'kit', value.name = 'median_genes')
predicted_u <- lapply(models_u, function(model) {
  predict(model$model, list(nreads=readrange))
}) |> as.data.table() |>
  mutate(nreads=readrange) |>
  melt(id.vars = 'nreads', variable.name = 'kit', value.name = 'median_umi')
```

## Extract RD50

```{r}
model_coef_g <- lapply(models_g, function(x) summary(x$model)$coefficients)
model_coef_u <- lapply(models_u, function(x) summary(x$model)$coefficients)

rd50_point_u <- lapply(names(model_coef_u), function(kit) {
  x <- model_coef_u[[kit]]['c', 'Estimate']
  y <- predict(models_u[[kit]]$model, list(nreads=x))
  return(list(Kit=kit, nreads=x, umi=y))
}) |> rbindlist()

rd50_point_g <- lapply(names(model_coef_g), function(kit) {
  x <- model_coef_g[[kit]]['c', 'Estimate']
  y <- predict(models_g[[kit]]$model, list(nreads=x))
  return(list(Kit=kit, nreads=x, genes=y))
}) |> rbindlist()

rd50_ribbon_g <- lapply(names(model_coef_g), function(kit) {
  x <- model_coef_g[[kit]]['c', 'Estimate']
  se <- model_coef_g[[kit]]['c', 'Std. Error']
  x <- seq(x-se,x+se, length.out=10)
  y <- predict(models_g[[kit]]$model, list(nreads=x))
  return(list(Kit=kit, x=x, y=y))
}) |> rbindlist()

rd50_ribbon_u <- lapply(names(model_coef_g), function(kit) {
  x <- model_coef_u[[kit]]['c', 'Estimate']
  se <- model_coef_u[[kit]]['c', 'Std. Error']
  x <- seq(x-se,x+se, length.out=10)
  y <- predict(models_u[[kit]]$model, list(nreads=x))
  return(list(Kit=kit, x=x, y=y))
}) |> rbindlist()

```

## Plot fit

```{r}
ggplot() +
  geom_line(data = predicted_g, aes(x=nreads, y=median_genes, color=kit)) +
  geom_point(data = filter(sat_curves, !is.na(median_genes)),
             aes(x=nreads, y=median_genes, color = Kit), size=.5) +
  geom_point(data = rd50_point_g, aes(x=nreads, y=genes, color=Kit), shape=108, size=9) +
  geom_ribbon(data = rd50_ribbon_g, aes(ymin=y-100, ymax=y+100, x=x, color=Kit, fill=Kit), linetype='dotted', alpha=0.25, show.legend = FALSE) +
  scale_color_manual(values = unlist(color_palette$kits), labels = label_function, na.translate = FALSE) +
  scale_fill_manual(values = unlist(color_palette$kits)) +
  guides(color = guide_legend(override.aes = list(size = 2))) +
  labs(x='Average reads per cell', y='Median genes per cell', color = 'Kit') 

```

# Plot

```{r}
ggplot(filter(sat_curves, !is.na(median_genes)),
       aes(x=nreads, y=median_genes, linetype = paste0(Individual, Replicate), color = Kit)) +
  geom_line() +
  geom_point(data = rd50_point_g, aes(x=nreads, y=genes, color=Kit),
             shape=43, size=14,
             inherit.aes = FALSE, show.legend = FALSE) +
  scale_color_manual(values = unlist(color_palette$kits), labels = label_function) +
  guides(color = guide_legend(override.aes = list(size = 2))) +
  labs(x='Average reads per cell', y='Median genes per cell', linetype='Sample', color = 'Kit') ->
  figures[['sat_curves_genes']]
  figures[['sat_curves_genes']]
```


```{r}
ggplot(filter(sat_curves, !is.na(median_umi)),
       aes(x=nreads, y=median_umi, linetype = paste0(Individual, Replicate), color = Kit)) +
  geom_line() +
  geom_point(data = rd50_point_u, aes(x=nreads, y=umi, color=Kit),
             shape=43, size=14,
             inherit.aes = FALSE, show.legend = FALSE) +
  scale_color_manual(values = unlist(color_palette$kits), labels = label_function, na.translate = FALSE) +
  guides(color = guide_legend(override.aes = list(size = 2))) +
  labs(x='Average reads per cell', y='Median transcripts per cell', linetype='Sample', color = 'Kit') ->
  figures[['sat_curves_umi']]
figures[['sat_curves_umi']]
```
